File: examples/simpleFit.cpp
// A simple fit in energy for signal and a background
//
// See/run util/make_simple_data to create some simple fake data to run this example.
#include <BinnedED.h>
#include <ROOTNtuple.h>
#include <BinnedNLLH.h>
#include <GridSearch.h>
#include <ParameterDict.h>

const std::string bgMCfile    = "";
const std::string sigMCfile   = "";
const std::string bgTreeName  = "";
const std::string sigTreeName = "";

const std::string dataFile = "";
const std::string dataTreeName = "";

int main(){
    ////////////////////
    // 1. Set Up PDFs //
    ////////////////////

    // Set up binning
    AxisCollection axes;
    axes.AddAxis(BinAxis("energy", 0, 10, 10, "Energy"));

    // Only interested in first bit of data ntuple
    ObsSet dataRep(0);

    // Set up pdf with these bins in this observable
    BinnedED bgPdf("bgPDF",axes);
    bgPdf.SetObservables(dataRep);
    BinnedED  signalPdf("signalPDF",axes);
    signalPdf.SetObservables(dataRep);

    std::cout << "Initialised Pdfs" << std::endl;

    /////////////////////////////////////
    // 2. Fill with data and normalise //
    /////////////////////////////////////

    ROOTNtuple bgMC(bgMCfile, bgTreeName);
    ROOTNtuple signalMC(sigMCfile, sigTreeName);

    for(size_t i = 0; i < bgMC.GetNEntries(); i++){
        bgPdf.Fill(bgMC.GetEntry(i));
    }

    for(size_t i = 0; i < signalMC.GetNEntries(); i++){
        signalPdf.Fill(signalMC.GetEntry(i));
    }

    bgPdf.Normalise();
    signalPdf.Normalise();

    std::cout << "Filled pdfs " << std::endl;

    ////////////////////////////
    // 3. Set Up LH function  //
    ////////////////////////////
    ROOTNtuple dataNt(dataFile, dataTreeName);
    BinnedNLLH lhFunction;
    lhFunction.SetDataSet(&dataNt); // initialise withe the data set
    lhFunction.AddDist(bgPdf);        
    lhFunction.AddDist(signalPdf);        
  
    std::cout << "Built LH function " << std::endl;

    // Set up the optimisation
    GridSearch gSearch;

    // Set up optimisation parameters.
    // Grid search needs a minimum, maximum and step size for each fit
    // parameter.
    // For BinnedEDs the normalisations are named by appending "_norm" to the
    // BinnedED name.
    ParameterDict minima;
    minima["bgPDF_norm"]= 800;
    minima["signalPDF_norm"]=800;

    ParameterDict maxima;
    maxima["bgPDF_norm"]= 2200;
    maxima["signalPDF_norm"]=2200;

    ParameterDict steps;
    steps["bgPDF_norm"]= 200;
    steps["signalPDF_norm"]=200;
    
    // Set optimisation parameters.
    gSearch.SetMinima(minima);
    gSearch.SetMaxima(maxima);
    gSearch.SetStepSizes(steps);

    ////////////
    // 4. Fit //
    ////////////
    FitResult result = gSearch.Optimise(&lhFunction);

    ParameterDict fit = result.GetBestFit();
    result.Print();
    result.SaveAs("simpleFit_result.txt");
    return 0;
}

File: examples/simple_fit.py
import argparse
import oxsx

if __name__=="__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-t", "--treename", type=str, default="output", help="Ntuple tree name")
    parser.add_argument("background", type=str, help="background filename")
    parser.add_argument("signal", type=str, help="signal filename")
    parser.add_argument("data", type=str, help="data filename")
    args = parser.parse_args()

    ###
    # Set up PDFs
    ###

    # Set up binning
    axes = oxsx.AxisCollection()
    axes.add_axis(oxsx.PdfAxis("energy", 2, 3, 10, "Energy"))

    # Only interested in first bit of data ntuple
    data_rep = oxsx.DataRepresentation(0)

    # Set up pdf with these bins in this observable
    bg_pdf = oxsx.BinnedPdf(axes)
    signal_pdf = oxsx.BinnedPdf(axes)
    bg_pdf.set_data_rep(data_rep)
    signal_pdf.set_data_rep(data_rep)

    print "Initialised Pdfs"

    ###
    # Fill with data and normalise
    ###

    bg_mc = oxsx.ROOTNtuple(args.background, args.treename)
    signal_mc = oxsx.ROOTNtuple(args.signal, args.treename)

    for i in range(0, bg_mc.get_n_entries()):
        bg_pdf.fill(bg_mc.get_entry(i))
    for i in range(0, signal_mc.get_n_entries()):
        signal_pdf.fill(signal_mc.get_entry(i))

    bg_pdf.normalise()
    signal_pdf.normalise()

    print "Filled Pdfs"

    ###
    # Set up LH function
    ###

    data_ntuple = oxsx.ROOTNtuple(args.data, args.treename)
    lh_function = oxsx.BinnedNLLH()
    lh_function.set_data_set(data_ntuple)
    lh_function.add_pdf(bg_pdf)
    lh_function.add_pdf(signal_pdf)

    print "Built LH function"

    search = oxsx.GridSearch()
    
    search.set_minima([1, 1])
    search.set_maxima([1000, 1000])
    search.set_step_sizes([1, 1])

    ###
    # Fit
    ###
    
    result = search.optimise(lh_function)

    fit = result.get_best_fit()
    result.print_result()
    result.save_as("simple_fit_result.txt")
